### PR TITLE
WIP: Delaying a validator set update by 1 block.

### PR DESCRIPTION
**Pull request by jaekwon**

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -74,7 +74,6 @@ type ConsensusState struct {
 	privValidator types.PrivValidator // for signing votes
 
 	// services for creating and executing blocks
-	// TODO: encapsulate all of this in one "BlockManager"
```
Comment (jaekwon) [inline consensus/state.go]: no managers plz.

File: state/state.go [hunk 2]
```diff
@@ -38,7 +38,8 @@ type State struct {
 	// so we can query for historical validator sets.
 	// Note that if s.LastBlockHeight causes a valset change,
 	// we set s.LastHeightValidatorsChanged = s.LastBlockHeight + 1
-	Validators                  *types.ValidatorSet
+	NextNextValidators          *types.ValidatorSet
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply state/state.go]: It's NextValidators and NextNextValidators, so as to distinguish between Validators and NextValidators.  The state stores validators responsible for signing the next block... so it seems consistent.  I'd be in favor of renaming ConsensusParams to NextConsensusParams as well, and AppHash to NextAppHash.  What do you think?

[the remark this answers is not in the document]
Comment (jaekwon) [reply state/state.go]: You're right, it should be Validators, esp since State is a field of ConsensusState & the way it's used there.

**Comment by kansi** [comment]
It would be helpful if you could point to the rationale behind this change?

**Comment by jaekwon** [comment]
Closing, replaced by https://github.com/tendermint/tendermint/pull/1756

**Pull request closed without merging**